// @aws-sdk/client-ses/dist/es/models/models_0.js
import { __assign } from "tslib";
export var AccountSendingPausedException;
(function (AccountSendingPausedException) {
    /**
     * @internal
     */
    AccountSendingPausedException.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(AccountSendingPausedException || (AccountSendingPausedException = {}));
export var AddHeaderAction;
(function (AddHeaderAction) {
    /**
     * @internal
     */
    AddHeaderAction.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(AddHeaderAction || (AddHeaderAction = {}));
export var AlreadyExistsException;
(function (AlreadyExistsException) {
    /**
     * @internal
     */
    AlreadyExistsException.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(AlreadyExistsException || (AlreadyExistsException = {}));
export var BehaviorOnMXFailure;
(function (BehaviorOnMXFailure) {
    BehaviorOnMXFailure["RejectMessage"] = "RejectMessage";
    BehaviorOnMXFailure["UseDefaultValue"] = "UseDefaultValue";
})(BehaviorOnMXFailure || (BehaviorOnMXFailure = {}));
export var Content;
(function (Content) {
    /**
     * @internal
     */
    Content.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(Content || (Content = {}));
export var Body;
(function (Body) {
    /**
     * @internal
     */
    Body.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(Body || (Body = {}));
export var BounceAction;
(function (BounceAction) {
    /**
     * @internal
     */
    BounceAction.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(BounceAction || (BounceAction = {}));
export var BounceType;
(function (BounceType) {
    BounceType["ContentRejected"] = "ContentRejected";
    BounceType["DoesNotExist"] = "DoesNotExist";
    BounceType["ExceededQuota"] = "ExceededQuota";
    BounceType["MessageTooLarge"] = "MessageTooLarge";
    BounceType["TemporaryFailure"] = "TemporaryFailure";
    BounceType["Undefined"] = "Undefined";
})(BounceType || (BounceType = {}));
export var DsnAction;
(function (DsnAction) {
    DsnAction["DELAYED"] = "delayed";
    DsnAction["DELIVERED"] = "delivered";
    DsnAction["EXPANDED"] = "expanded";
    DsnAction["FAILED"] = "failed";
    DsnAction["RELAYED"] = "relayed";
})(DsnAction || (DsnAction = {}));
export var ExtensionField;
(function (ExtensionField) {
    /**
     * @internal
     */
    ExtensionField.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(ExtensionField || (ExtensionField = {}));
export var RecipientDsnFields;
(function (RecipientDsnFields) {
    /**
     * @internal
     */
    RecipientDsnFields.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(RecipientDsnFields || (RecipientDsnFields = {}));
export var BouncedRecipientInfo;
(function (BouncedRecipientInfo) {
    /**
     * @internal
     */
    BouncedRecipientInfo.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(BouncedRecipientInfo || (BouncedRecipientInfo = {}));
export var Destination;
(function (Destination) {
    /**
     * @internal
     */
    Destination.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(Destination || (Destination = {}));
export var MessageTag;
(function (MessageTag) {
    /**
     * @internal
     */
    MessageTag.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(MessageTag || (MessageTag = {}));
export var BulkEmailDestination;
(function (BulkEmailDestination) {
    /**
     * @internal
     */
    BulkEmailDestination.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(BulkEmailDestination || (BulkEmailDestination = {}));
export var BulkEmailStatus;
(function (BulkEmailStatus) {
    BulkEmailStatus["AccountDailyQuotaExceeded"] = "AccountDailyQuotaExceeded";
    BulkEmailStatus["AccountSendingPaused"] = "AccountSendingPaused";
    BulkEmailStatus["AccountSuspended"] = "AccountSuspended";
    BulkEmailStatus["AccountThrottled"] = "AccountThrottled";
    BulkEmailStatus["ConfigurationSetDoesNotExist"] = "ConfigurationSetDoesNotExist";
    BulkEmailStatus["ConfigurationSetSendingPaused"] = "ConfigurationSetSendingPaused";
    BulkEmailStatus["Failed"] = "Failed";
    BulkEmailStatus["InvalidParameterValue"] = "InvalidParameterValue";
    BulkEmailStatus["InvalidSendingPoolName"] = "InvalidSendingPoolName";
    BulkEmailStatus["MailFromDomainNotVerified"] = "MailFromDomainNotVerified";
    BulkEmailStatus["MessageRejected"] = "MessageRejected";
    BulkEmailStatus["Success"] = "Success";
    BulkEmailStatus["TemplateDoesNotExist"] = "TemplateDoesNotExist";
    BulkEmailStatus["TransientFailure"] = "TransientFailure";
})(BulkEmailStatus || (BulkEmailStatus = {}));
export var BulkEmailDestinationStatus;
(function (BulkEmailDestinationStatus) {
    /**
     * @internal
     */
    BulkEmailDestinationStatus.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(BulkEmailDestinationStatus || (BulkEmailDestinationStatus = {}));
export var CannotDeleteException;
(function (CannotDeleteException) {
    /**
     * @internal
     */
    CannotDeleteException.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(CannotDeleteException || (CannotDeleteException = {}));
export var CloneReceiptRuleSetRequest;
(function (CloneReceiptRuleSetRequest) {
    /**
     * @internal
     */
    CloneReceiptRuleSetRequest.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(CloneReceiptRuleSetRequest || (CloneReceiptRuleSetRequest = {}));
export var CloneReceiptRuleSetResponse;
(function (CloneReceiptRuleSetResponse) {
    /**
     * @internal
     */
    CloneReceiptRuleSetResponse.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(CloneReceiptRuleSetResponse || (CloneReceiptRuleSetResponse = {}));
export var LimitExceededException;
(function (LimitExceededException) {
    /**
     * @internal
     */
    LimitExceededException.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(LimitExceededException || (LimitExceededException = {}));
export var RuleSetDoesNotExistException;
(function (RuleSetDoesNotExistException) {
    /**
     * @internal
     */
    RuleSetDoesNotExistException.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(RuleSetDoesNotExistException || (RuleSetDoesNotExistException = {}));
export var DimensionValueSource;
(function (DimensionValueSource) {
    DimensionValueSource["EMAIL_HEADER"] = "emailHeader";
    DimensionValueSource["LINK_TAG"] = "linkTag";
    DimensionValueSource["MESSAGE_TAG"] = "messageTag";
})(DimensionValueSource || (DimensionValueSource = {}));
export var CloudWatchDimensionConfiguration;
(function (CloudWatchDimensionConfiguration) {
    /**
     * @internal
     */
    CloudWatchDimensionConfiguration.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(CloudWatchDimensionConfiguration || (CloudWatchDimensionConfiguration = {}));
export var CloudWatchDestination;
(function (CloudWatchDestination) {
    /**
     * @internal
     */
    CloudWatchDestination.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(CloudWatchDestination || (CloudWatchDestination = {}));
export var ConfigurationSet;
(function (ConfigurationSet) {
    /**
     * @internal
     */
    ConfigurationSet.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(ConfigurationSet || (ConfigurationSet = {}));
export var ConfigurationSetAlreadyExistsException;
(function (ConfigurationSetAlreadyExistsException) {
    /**
     * @internal
     */
    ConfigurationSetAlreadyExistsException.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(ConfigurationSetAlreadyExistsException || (ConfigurationSetAlreadyExistsException = {}));
export var ConfigurationSetAttribute;
(function (ConfigurationSetAttribute) {
    ConfigurationSetAttribute["DELIVERY_OPTIONS"] = "deliveryOptions";
    ConfigurationSetAttribute["EVENT_DESTINATIONS"] = "eventDestinations";
    ConfigurationSetAttribute["REPUTATION_OPTIONS"] = "reputationOptions";
    ConfigurationSetAttribute["TRACKING_OPTIONS"] = "trackingOptions";
})(ConfigurationSetAttribute || (ConfigurationSetAttribute = {}));
export var ConfigurationSetDoesNotExistException;
(function (ConfigurationSetDoesNotExistException) {
    /**
     * @internal
     */
    ConfigurationSetDoesNotExistException.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(ConfigurationSetDoesNotExistException || (ConfigurationSetDoesNotExistException = {}));
export var ConfigurationSetSendingPausedException;
(function (ConfigurationSetSendingPausedException) {
    /**
     * @internal
     */
    ConfigurationSetSendingPausedException.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(ConfigurationSetSendingPausedException || (ConfigurationSetSendingPausedException = {}));
export var CreateConfigurationSetRequest;
(function (CreateConfigurationSetRequest) {
    /**
     * @internal
     */
    CreateConfigurationSetRequest.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(CreateConfigurationSetRequest || (CreateConfigurationSetRequest = {}));
export var CreateConfigurationSetResponse;
(function (CreateConfigurationSetResponse) {
    /**
     * @internal
     */
    CreateConfigurationSetResponse.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(CreateConfigurationSetResponse || (CreateConfigurationSetResponse = {}));
export var InvalidConfigurationSetException;
(function (InvalidConfigurationSetException) {
    /**
     * @internal
     */
    InvalidConfigurationSetException.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(InvalidConfigurationSetException || (InvalidConfigurationSetException = {}));
export var KinesisFirehoseDestination;
(function (KinesisFirehoseDestination) {
    /**
     * @internal
     */
    KinesisFirehoseDestination.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(KinesisFirehoseDestination || (KinesisFirehoseDestination = {}));
export var EventType;
(function (EventType) {
    EventType["BOUNCE"] = "bounce";
    EventType["CLICK"] = "click";
    EventType["COMPLAINT"] = "complaint";
    EventType["DELIVERY"] = "delivery";
    EventType["OPEN"] = "open";
    EventType["REJECT"] = "reject";
    EventType["RENDERING_FAILURE"] = "renderingFailure";
    EventType["SEND"] = "send";
})(EventType || (EventType = {}));
export var SNSDestination;
(function (SNSDestination) {
    /**
     * @internal
     */
    SNSDestination.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(SNSDestination || (SNSDestination = {}));
export var EventDestination;
(function (EventDestination) {
    /**
     * @internal
     */
    EventDestination.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(EventDestination || (EventDestination = {}));
export var CreateConfigurationSetEventDestinationRequest;
(function (CreateConfigurationSetEventDestinationRequest) {
    /**
     * @internal
     */
    CreateConfigurationSetEventDestinationRequest.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(CreateConfigurationSetEventDestinationRequest || (CreateConfigurationSetEventDestinationRequest = {}));
export var CreateConfigurationSetEventDestinationResponse;
(function (CreateConfigurationSetEventDestinationResponse) {
    /**
     * @internal
     */
    CreateConfigurationSetEventDestinationResponse.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(CreateConfigurationSetEventDestinationResponse || (CreateConfigurationSetEventDestinationResponse = {}));
export var EventDestinationAlreadyExistsException;
(function (EventDestinationAlreadyExistsException) {
    /**
     * @internal
     */
    EventDestinationAlreadyExistsException.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(EventDestinationAlreadyExistsException || (EventDestinationAlreadyExistsException = {}));
export var InvalidCloudWatchDestinationException;
(function (InvalidCloudWatchDestinationException) {
    /**
     * @internal
     */
    InvalidCloudWatchDestinationException.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(InvalidCloudWatchDestinationException || (InvalidCloudWatchDestinationException = {}));
export var InvalidFirehoseDestinationException;
(function (InvalidFirehoseDestinationException) {
    /**
     * @internal
     */
    InvalidFirehoseDestinationException.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(InvalidFirehoseDestinationException || (InvalidFirehoseDestinationException = {}));
export var InvalidSNSDestinationException;
(function (InvalidSNSDestinationException) {
    /**
     * @internal
     */
    InvalidSNSDestinationException.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(InvalidSNSDestinationException || (InvalidSNSDestinationException = {}));
export var TrackingOptions;
(function (TrackingOptions) {
    /**
     * @internal
     */
    TrackingOptions.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(TrackingOptions || (TrackingOptions = {}));
export var CreateConfigurationSetTrackingOptionsRequest;
(function (CreateConfigurationSetTrackingOptionsRequest) {
    /**
     * @internal
     */
    CreateConfigurationSetTrackingOptionsRequest.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(CreateConfigurationSetTrackingOptionsRequest || (CreateConfigurationSetTrackingOptionsRequest = {}));
export var CreateConfigurationSetTrackingOptionsResponse;
(function (CreateConfigurationSetTrackingOptionsResponse) {
    /**
     * @internal
     */
    CreateConfigurationSetTrackingOptionsResponse.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(CreateConfigurationSetTrackingOptionsResponse || (CreateConfigurationSetTrackingOptionsResponse = {}));
export var InvalidTrackingOptionsException;
(function (InvalidTrackingOptionsException) {
    /**
     * @internal
     */
    InvalidTrackingOptionsException.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(InvalidTrackingOptionsException || (InvalidTrackingOptionsException = {}));
export var TrackingOptionsAlreadyExistsException;
(function (TrackingOptionsAlreadyExistsException) {
    /**
     * @internal
     */
    TrackingOptionsAlreadyExistsException.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(TrackingOptionsAlreadyExistsException || (TrackingOptionsAlreadyExistsException = {}));
export var CreateCustomVerificationEmailTemplateRequest;
(function (CreateCustomVerificationEmailTemplateRequest) {
    /**
     * @internal
     */
    CreateCustomVerificationEmailTemplateRequest.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(CreateCustomVerificationEmailTemplateRequest || (CreateCustomVerificationEmailTemplateRequest = {}));
export var CustomVerificationEmailInvalidContentException;
(function (CustomVerificationEmailInvalidContentException) {
    /**
     * @internal
     */
    CustomVerificationEmailInvalidContentException.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(CustomVerificationEmailInvalidContentException || (CustomVerificationEmailInvalidContentException = {}));
export var CustomVerificationEmailTemplateAlreadyExistsException;
(function (CustomVerificationEmailTemplateAlreadyExistsException) {
    /**
     * @internal
     */
    CustomVerificationEmailTemplateAlreadyExistsException.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(CustomVerificationEmailTemplateAlreadyExistsException || (CustomVerificationEmailTemplateAlreadyExistsException = {}));
export var FromEmailAddressNotVerifiedException;
(function (FromEmailAddressNotVerifiedException) {
    /**
     * @internal
     */
    FromEmailAddressNotVerifiedException.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(FromEmailAddressNotVerifiedException || (FromEmailAddressNotVerifiedException = {}));
export var ReceiptFilterPolicy;
(function (ReceiptFilterPolicy) {
    ReceiptFilterPolicy["Allow"] = "Allow";
    ReceiptFilterPolicy["Block"] = "Block";
})(ReceiptFilterPolicy || (ReceiptFilterPolicy = {}));
export var ReceiptIpFilter;
(function (ReceiptIpFilter) {
    /**
     * @internal
     */
    ReceiptIpFilter.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(ReceiptIpFilter || (ReceiptIpFilter = {}));
export var ReceiptFilter;
(function (ReceiptFilter) {
    /**
     * @internal
     */
    ReceiptFilter.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(ReceiptFilter || (ReceiptFilter = {}));
export var CreateReceiptFilterRequest;
(function (CreateReceiptFilterRequest) {
    /**
     * @internal
     */
    CreateReceiptFilterRequest.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(CreateReceiptFilterRequest || (CreateReceiptFilterRequest = {}));
export var CreateReceiptFilterResponse;
(function (CreateReceiptFilterResponse) {
    /**
     * @internal
     */
    CreateReceiptFilterResponse.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(CreateReceiptFilterResponse || (CreateReceiptFilterResponse = {}));
export var LambdaAction;
(function (LambdaAction) {
    /**
     * @internal
     */
    LambdaAction.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(LambdaAction || (LambdaAction = {}));
export var S3Action;
(function (S3Action) {
    /**
     * @internal
     */
    S3Action.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(S3Action || (S3Action = {}));
export var SNSActionEncoding;
(function (SNSActionEncoding) {
    SNSActionEncoding["Base64"] = "Base64";
    SNSActionEncoding["UTF8"] = "UTF-8";
})(SNSActionEncoding || (SNSActionEncoding = {}));
export var SNSAction;
(function (SNSAction) {
    /**
     * @internal
     */
    SNSAction.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(SNSAction || (SNSAction = {}));
export var StopScope;
(function (StopScope) {
    StopScope["RULE_SET"] = "RuleSet";
})(StopScope || (StopScope = {}));
export var StopAction;
(function (StopAction) {
    /**
     * @internal
     */
    StopAction.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(StopAction || (StopAction = {}));
export var WorkmailAction;
(function (WorkmailAction) {
    /**
     * @internal
     */
    WorkmailAction.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(WorkmailAction || (WorkmailAction = {}));
export var ReceiptAction;
(function (ReceiptAction) {
    /**
     * @internal
     */
    ReceiptAction.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(ReceiptAction || (ReceiptAction = {}));
export var TlsPolicy;
(function (TlsPolicy) {
    TlsPolicy["Optional"] = "Optional";
    TlsPolicy["Require"] = "Require";
})(TlsPolicy || (TlsPolicy = {}));
export var ReceiptRule;
(function (ReceiptRule) {
    /**
     * @internal
     */
    ReceiptRule.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(ReceiptRule || (ReceiptRule = {}));
export var CreateReceiptRuleRequest;
(function (CreateReceiptRuleRequest) {
    /**
     * @internal
     */
    CreateReceiptRuleRequest.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(CreateReceiptRuleRequest || (CreateReceiptRuleRequest = {}));
export var CreateReceiptRuleResponse;
(function (CreateReceiptRuleResponse) {
    /**
     * @internal
     */
    CreateReceiptRuleResponse.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(CreateReceiptRuleResponse || (CreateReceiptRuleResponse = {}));
export var InvalidLambdaFunctionException;
(function (InvalidLambdaFunctionException) {
    /**
     * @internal
     */
    InvalidLambdaFunctionException.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(InvalidLambdaFunctionException || (InvalidLambdaFunctionException = {}));
export var InvalidS3ConfigurationException;
(function (InvalidS3ConfigurationException) {
    /**
     * @internal
     */
    InvalidS3ConfigurationException.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(InvalidS3ConfigurationException || (InvalidS3ConfigurationException = {}));
export var InvalidSnsTopicException;
(function (InvalidSnsTopicException) {
    /**
     * @internal
     */
    InvalidSnsTopicException.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(InvalidSnsTopicException || (InvalidSnsTopicException = {}));
export var RuleDoesNotExistException;
(function (RuleDoesNotExistException) {
    /**
     * @internal
     */
    RuleDoesNotExistException.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(RuleDoesNotExistException || (RuleDoesNotExistException = {}));
export var CreateReceiptRuleSetRequest;
(function (CreateReceiptRuleSetRequest) {
    /**
     * @internal
     */
    CreateReceiptRuleSetRequest.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(CreateReceiptRuleSetRequest || (CreateReceiptRuleSetRequest = {}));
export var CreateReceiptRuleSetResponse;
(function (CreateReceiptRuleSetResponse) {
    /**
     * @internal
     */
    CreateReceiptRuleSetResponse.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(CreateReceiptRuleSetResponse || (CreateReceiptRuleSetResponse = {}));
export var Template;
(function (Template) {
    /**
     * @internal
     */
    Template.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(Template || (Template = {}));
export var CreateTemplateRequest;
(function (CreateTemplateRequest) {
    /**
     * @internal
     */
    CreateTemplateRequest.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(CreateTemplateRequest || (CreateTemplateRequest = {}));
export var CreateTemplateResponse;
(function (CreateTemplateResponse) {
    /**
     * @internal
     */
    CreateTemplateResponse.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(CreateTemplateResponse || (CreateTemplateResponse = {}));
export var InvalidTemplateException;
(function (InvalidTemplateException) {
    /**
     * @internal
     */
    InvalidTemplateException.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(InvalidTemplateException || (InvalidTemplateException = {}));
export var CustomMailFromStatus;
(function (CustomMailFromStatus) {
    CustomMailFromStatus["Failed"] = "Failed";
    CustomMailFromStatus["Pending"] = "Pending";
    CustomMailFromStatus["Success"] = "Success";
    CustomMailFromStatus["TemporaryFailure"] = "TemporaryFailure";
})(CustomMailFromStatus || (CustomMailFromStatus = {}));
export var CustomVerificationEmailTemplate;
(function (CustomVerificationEmailTemplate) {
    /**
     * @internal
     */
    CustomVerificationEmailTemplate.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(CustomVerificationEmailTemplate || (CustomVerificationEmailTemplate = {}));
export var CustomVerificationEmailTemplateDoesNotExistException;
(function (CustomVerificationEmailTemplateDoesNotExistException) {
    /**
     * @internal
     */
    CustomVerificationEmailTemplateDoesNotExistException.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(CustomVerificationEmailTemplateDoesNotExistException || (CustomVerificationEmailTemplateDoesNotExistException = {}));
export var DeleteConfigurationSetRequest;
(function (DeleteConfigurationSetRequest) {
    /**
     * @internal
     */
    DeleteConfigurationSetRequest.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(DeleteConfigurationSetRequest || (DeleteConfigurationSetRequest = {}));
export var DeleteConfigurationSetResponse;
(function (DeleteConfigurationSetResponse) {
    /**
     * @internal
     */
    DeleteConfigurationSetResponse.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(DeleteConfigurationSetResponse || (DeleteConfigurationSetResponse = {}));
export var DeleteConfigurationSetEventDestinationRequest;
(function (DeleteConfigurationSetEventDestinationRequest) {
    /**
     * @internal
     */
    DeleteConfigurationSetEventDestinationRequest.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(DeleteConfigurationSetEventDestinationRequest || (DeleteConfigurationSetEventDestinationRequest = {}));
export var DeleteConfigurationSetEventDestinationResponse;
(function (DeleteConfigurationSetEventDestinationResponse) {
    /**
     * @internal
     */
    DeleteConfigurationSetEventDestinationResponse.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(DeleteConfigurationSetEventDestinationResponse || (DeleteConfigurationSetEventDestinationResponse = {}));
export var EventDestinationDoesNotExistException;
(function (EventDestinationDoesNotExistException) {
    /**
     * @internal
     */
    EventDestinationDoesNotExistException.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(EventDestinationDoesNotExistException || (EventDestinationDoesNotExistException = {}));
export var DeleteConfigurationSetTrackingOptionsRequest;
(function (DeleteConfigurationSetTrackingOptionsRequest) {
    /**
     * @internal
     */
    DeleteConfigurationSetTrackingOptionsRequest.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(DeleteConfigurationSetTrackingOptionsRequest || (DeleteConfigurationSetTrackingOptionsRequest = {}));
export var DeleteConfigurationSetTrackingOptionsResponse;
(function (DeleteConfigurationSetTrackingOptionsResponse) {
    /**
     * @internal
     */
    DeleteConfigurationSetTrackingOptionsResponse.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(DeleteConfigurationSetTrackingOptionsResponse || (DeleteConfigurationSetTrackingOptionsResponse = {}));
export var TrackingOptionsDoesNotExistException;
(function (TrackingOptionsDoesNotExistException) {
    /**
     * @internal
     */
    TrackingOptionsDoesNotExistException.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(TrackingOptionsDoesNotExistException || (TrackingOptionsDoesNotExistException = {}));
export var DeleteCustomVerificationEmailTemplateRequest;
(function (DeleteCustomVerificationEmailTemplateRequest) {
    /**
     * @internal
     */
    DeleteCustomVerificationEmailTemplateRequest.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(DeleteCustomVerificationEmailTemplateRequest || (DeleteCustomVerificationEmailTemplateRequest = {}));
export var DeleteIdentityRequest;
(function (DeleteIdentityRequest) {
    /**
     * @internal
     */
    DeleteIdentityRequest.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(DeleteIdentityRequest || (DeleteIdentityRequest = {}));
export var DeleteIdentityResponse;
(function (DeleteIdentityResponse) {
    /**
     * @internal
     */
    DeleteIdentityResponse.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(DeleteIdentityResponse || (DeleteIdentityResponse = {}));
export var DeleteIdentityPolicyRequest;
(function (DeleteIdentityPolicyRequest) {
    /**
     * @internal
     */
    DeleteIdentityPolicyRequest.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(DeleteIdentityPolicyRequest || (DeleteIdentityPolicyRequest = {}));
export var DeleteIdentityPolicyResponse;
(function (DeleteIdentityPolicyResponse) {
    /**
     * @internal
     */
    DeleteIdentityPolicyResponse.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(DeleteIdentityPolicyResponse || (DeleteIdentityPolicyResponse = {}));
export var DeleteReceiptFilterRequest;
(function (DeleteReceiptFilterRequest) {
    /**
     * @internal
     */
    DeleteReceiptFilterRequest.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(DeleteReceiptFilterRequest || (DeleteReceiptFilterRequest = {}));
export var DeleteReceiptFilterResponse;
(function (DeleteReceiptFilterResponse) {
    /**
     * @internal
     */
    DeleteReceiptFilterResponse.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(DeleteReceiptFilterResponse || (DeleteReceiptFilterResponse = {}));
export var DeleteReceiptRuleRequest;
(function (DeleteReceiptRuleRequest) {
    /**
     * @internal
     */
    DeleteReceiptRuleRequest.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(DeleteReceiptRuleRequest || (DeleteReceiptRuleRequest = {}));
export var DeleteReceiptRuleResponse;
(function (DeleteReceiptRuleResponse) {
    /**
     * @internal
     */
    DeleteReceiptRuleResponse.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(DeleteReceiptRuleResponse || (DeleteReceiptRuleResponse = {}));
export var DeleteReceiptRuleSetRequest;
(function (DeleteReceiptRuleSetRequest) {
    /**
     * @internal
     */
    DeleteReceiptRuleSetRequest.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(DeleteReceiptRuleSetRequest || (DeleteReceiptRuleSetRequest = {}));
export var DeleteReceiptRuleSetResponse;
(function (DeleteReceiptRuleSetResponse) {
    /**
     * @internal
     */
    DeleteReceiptRuleSetResponse.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(DeleteReceiptRuleSetResponse || (DeleteReceiptRuleSetResponse = {}));
export var DeleteTemplateRequest;
(function (DeleteTemplateRequest) {
    /**
     * @internal
     */
    DeleteTemplateRequest.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(DeleteTemplateRequest || (DeleteTemplateRequest = {}));
export var DeleteTemplateResponse;
(function (DeleteTemplateResponse) {
    /**
     * @internal
     */
    DeleteTemplateResponse.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(DeleteTemplateResponse || (DeleteTemplateResponse = {}));
export var DeleteVerifiedEmailAddressRequest;
(function (DeleteVerifiedEmailAddressRequest) {
    /**
     * @internal
     */
    DeleteVerifiedEmailAddressRequest.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(DeleteVerifiedEmailAddressRequest || (DeleteVerifiedEmailAddressRequest = {}));
export var DeliveryOptions;
(function (DeliveryOptions) {
    /**
     * @internal
     */
    DeliveryOptions.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(DeliveryOptions || (DeliveryOptions = {}));
export var DescribeActiveReceiptRuleSetRequest;
(function (DescribeActiveReceiptRuleSetRequest) {
    /**
     * @internal
     */
    DescribeActiveReceiptRuleSetRequest.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(DescribeActiveReceiptRuleSetRequest || (DescribeActiveReceiptRuleSetRequest = {}));
export var ReceiptRuleSetMetadata;
(function (ReceiptRuleSetMetadata) {
    /**
     * @internal
     */
    ReceiptRuleSetMetadata.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(ReceiptRuleSetMetadata || (ReceiptRuleSetMetadata = {}));
export var DescribeActiveReceiptRuleSetResponse;
(function (DescribeActiveReceiptRuleSetResponse) {
    /**
     * @internal
     */
    DescribeActiveReceiptRuleSetResponse.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(DescribeActiveReceiptRuleSetResponse || (DescribeActiveReceiptRuleSetResponse = {}));
export var DescribeConfigurationSetRequest;
(function (DescribeConfigurationSetRequest) {
    /**
     * @internal
     */
    DescribeConfigurationSetRequest.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(DescribeConfigurationSetRequest || (DescribeConfigurationSetRequest = {}));
export var ReputationOptions;
(function (ReputationOptions) {
    /**
     * @internal
     */
    ReputationOptions.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(ReputationOptions || (ReputationOptions = {}));
export var DescribeConfigurationSetResponse;
(function (DescribeConfigurationSetResponse) {
    /**
     * @internal
     */
    DescribeConfigurationSetResponse.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(DescribeConfigurationSetResponse || (DescribeConfigurationSetResponse = {}));
export var DescribeReceiptRuleRequest;
(function (DescribeReceiptRuleRequest) {
    /**
     * @internal
     */
    DescribeReceiptRuleRequest.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(DescribeReceiptRuleRequest || (DescribeReceiptRuleRequest = {}));
export var DescribeReceiptRuleResponse;
(function (DescribeReceiptRuleResponse) {
    /**
     * @internal
     */
    DescribeReceiptRuleResponse.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(DescribeReceiptRuleResponse || (DescribeReceiptRuleResponse = {}));
export var DescribeReceiptRuleSetRequest;
(function (DescribeReceiptRuleSetRequest) {
    /**
     * @internal
     */
    DescribeReceiptRuleSetRequest.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(DescribeReceiptRuleSetRequest || (DescribeReceiptRuleSetRequest = {}));
export var DescribeReceiptRuleSetResponse;
(function (DescribeReceiptRuleSetResponse) {
    /**
     * @internal
     */
    DescribeReceiptRuleSetResponse.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(DescribeReceiptRuleSetResponse || (DescribeReceiptRuleSetResponse = {}));
export var IdentityDkimAttributes;
(function (IdentityDkimAttributes) {
    /**
     * @internal
     */
    IdentityDkimAttributes.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(IdentityDkimAttributes || (IdentityDkimAttributes = {}));
export var GetAccountSendingEnabledResponse;
(function (GetAccountSendingEnabledResponse) {
    /**
     * @internal
     */
    GetAccountSendingEnabledResponse.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(GetAccountSendingEnabledResponse || (GetAccountSendingEnabledResponse = {}));
export var GetCustomVerificationEmailTemplateRequest;
(function (GetCustomVerificationEmailTemplateRequest) {
    /**
     * @internal
     */
    GetCustomVerificationEmailTemplateRequest.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(GetCustomVerificationEmailTemplateRequest || (GetCustomVerificationEmailTemplateRequest = {}));
export var GetCustomVerificationEmailTemplateResponse;
(function (GetCustomVerificationEmailTemplateResponse) {
    /**
     * @internal
     */
    GetCustomVerificationEmailTemplateResponse.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(GetCustomVerificationEmailTemplateResponse || (GetCustomVerificationEmailTemplateResponse = {}));
export var GetIdentityDkimAttributesRequest;
(function (GetIdentityDkimAttributesRequest) {
    /**
     * @internal
     */
    GetIdentityDkimAttributesRequest.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(GetIdentityDkimAttributesRequest || (GetIdentityDkimAttributesRequest = {}));
export var GetIdentityDkimAttributesResponse;
(function (GetIdentityDkimAttributesResponse) {
    /**
     * @internal
     */
    GetIdentityDkimAttributesResponse.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(GetIdentityDkimAttributesResponse || (GetIdentityDkimAttributesResponse = {}));
export var GetIdentityMailFromDomainAttributesRequest;
(function (GetIdentityMailFromDomainAttributesRequest) {
    /**
     * @internal
     */
    GetIdentityMailFromDomainAttributesRequest.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(GetIdentityMailFromDomainAttributesRequest || (GetIdentityMailFromDomainAttributesRequest = {}));
export var IdentityMailFromDomainAttributes;
(function (IdentityMailFromDomainAttributes) {
    /**
     * @internal
     */
    IdentityMailFromDomainAttributes.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(IdentityMailFromDomainAttributes || (IdentityMailFromDomainAttributes = {}));
export var GetIdentityMailFromDomainAttributesResponse;
(function (GetIdentityMailFromDomainAttributesResponse) {
    /**
     * @internal
     */
    GetIdentityMailFromDomainAttributesResponse.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(GetIdentityMailFromDomainAttributesResponse || (GetIdentityMailFromDomainAttributesResponse = {}));
export var GetIdentityNotificationAttributesRequest;
(function (GetIdentityNotificationAttributesRequest) {
    /**
     * @internal
     */
    GetIdentityNotificationAttributesRequest.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(GetIdentityNotificationAttributesRequest || (GetIdentityNotificationAttributesRequest = {}));
export var IdentityNotificationAttributes;
(function (IdentityNotificationAttributes) {
    /**
     * @internal
     */
    IdentityNotificationAttributes.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(IdentityNotificationAttributes || (IdentityNotificationAttributes = {}));
export var GetIdentityNotificationAttributesResponse;
(function (GetIdentityNotificationAttributesResponse) {
    /**
     * @internal
     */
    GetIdentityNotificationAttributesResponse.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(GetIdentityNotificationAttributesResponse || (GetIdentityNotificationAttributesResponse = {}));
export var GetIdentityPoliciesRequest;
(function (GetIdentityPoliciesRequest) {
    /**
     * @internal
     */
    GetIdentityPoliciesRequest.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(GetIdentityPoliciesRequest || (GetIdentityPoliciesRequest = {}));
export var GetIdentityPoliciesResponse;
(function (GetIdentityPoliciesResponse) {
    /**
     * @internal
     */
    GetIdentityPoliciesResponse.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(GetIdentityPoliciesResponse || (GetIdentityPoliciesResponse = {}));
export var GetIdentityVerificationAttributesRequest;
(function (GetIdentityVerificationAttributesRequest) {
    /**
     * @internal
     */
    GetIdentityVerificationAttributesRequest.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(GetIdentityVerificationAttributesRequest || (GetIdentityVerificationAttributesRequest = {}));
export var IdentityVerificationAttributes;
(function (IdentityVerificationAttributes) {
    /**
     * @internal
     */
    IdentityVerificationAttributes.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(IdentityVerificationAttributes || (IdentityVerificationAttributes = {}));
export var GetIdentityVerificationAttributesResponse;
(function (GetIdentityVerificationAttributesResponse) {
    /**
     * @internal
     */
    GetIdentityVerificationAttributesResponse.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(GetIdentityVerificationAttributesResponse || (GetIdentityVerificationAttributesResponse = {}));
export var GetSendQuotaResponse;
(function (GetSendQuotaResponse) {
    /**
     * @internal
     */
    GetSendQuotaResponse.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(GetSendQuotaResponse || (GetSendQuotaResponse = {}));
export var SendDataPoint;
(function (SendDataPoint) {
    /**
     * @internal
     */
    SendDataPoint.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(SendDataPoint || (SendDataPoint = {}));
export var GetSendStatisticsResponse;
(function (GetSendStatisticsResponse) {
    /**
     * @internal
     */
    GetSendStatisticsResponse.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(GetSendStatisticsResponse || (GetSendStatisticsResponse = {}));
export var GetTemplateRequest;
(function (GetTemplateRequest) {
    /**
     * @internal
     */
    GetTemplateRequest.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(GetTemplateRequest || (GetTemplateRequest = {}));
export var GetTemplateResponse;
(function (GetTemplateResponse) {
    /**
     * @internal
     */
    GetTemplateResponse.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(GetTemplateResponse || (GetTemplateResponse = {}));
export var TemplateDoesNotExistException;
(function (TemplateDoesNotExistException) {
    /**
     * @internal
     */
    TemplateDoesNotExistException.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(TemplateDoesNotExistException || (TemplateDoesNotExistException = {}));
export var InvalidDeliveryOptionsException;
(function (InvalidDeliveryOptionsException) {
    /**
     * @internal
     */
    InvalidDeliveryOptionsException.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(InvalidDeliveryOptionsException || (InvalidDeliveryOptionsException = {}));
export var InvalidPolicyException;
(function (InvalidPolicyException) {
    /**
     * @internal
     */
    InvalidPolicyException.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(InvalidPolicyException || (InvalidPolicyException = {}));
export var InvalidRenderingParameterException;
(function (InvalidRenderingParameterException) {
    /**
     * @internal
     */
    InvalidRenderingParameterException.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(InvalidRenderingParameterException || (InvalidRenderingParameterException = {}));
export var ListConfigurationSetsRequest;
(function (ListConfigurationSetsRequest) {
    /**
     * @internal
     */
    ListConfigurationSetsRequest.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(ListConfigurationSetsRequest || (ListConfigurationSetsRequest = {}));
export var ListConfigurationSetsResponse;
(function (ListConfigurationSetsResponse) {
    /**
     * @internal
     */
    ListConfigurationSetsResponse.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(ListConfigurationSetsResponse || (ListConfigurationSetsResponse = {}));
export var ListCustomVerificationEmailTemplatesRequest;
(function (ListCustomVerificationEmailTemplatesRequest) {
    /**
     * @internal
     */
    ListCustomVerificationEmailTemplatesRequest.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(ListCustomVerificationEmailTemplatesRequest || (ListCustomVerificationEmailTemplatesRequest = {}));
export var ListCustomVerificationEmailTemplatesResponse;
(function (ListCustomVerificationEmailTemplatesResponse) {
    /**
     * @internal
     */
    ListCustomVerificationEmailTemplatesResponse.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(ListCustomVerificationEmailTemplatesResponse || (ListCustomVerificationEmailTemplatesResponse = {}));
export var ListIdentitiesRequest;
(function (ListIdentitiesRequest) {
    /**
     * @internal
     */
    ListIdentitiesRequest.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(ListIdentitiesRequest || (ListIdentitiesRequest = {}));
export var ListIdentitiesResponse;
(function (ListIdentitiesResponse) {
    /**
     * @internal
     */
    ListIdentitiesResponse.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(ListIdentitiesResponse || (ListIdentitiesResponse = {}));
export var ListIdentityPoliciesRequest;
(function (ListIdentityPoliciesRequest) {
    /**
     * @internal
     */
    ListIdentityPoliciesRequest.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(ListIdentityPoliciesRequest || (ListIdentityPoliciesRequest = {}));
export var ListIdentityPoliciesResponse;
(function (ListIdentityPoliciesResponse) {
    /**
     * @internal
     */
    ListIdentityPoliciesResponse.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(ListIdentityPoliciesResponse || (ListIdentityPoliciesResponse = {}));
export var ListReceiptFiltersRequest;
(function (ListReceiptFiltersRequest) {
    /**
     * @internal
     */
    ListReceiptFiltersRequest.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(ListReceiptFiltersRequest || (ListReceiptFiltersRequest = {}));
export var ListReceiptFiltersResponse;
(function (ListReceiptFiltersResponse) {
    /**
     * @internal
     */
    ListReceiptFiltersResponse.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(ListReceiptFiltersResponse || (ListReceiptFiltersResponse = {}));
export var ListReceiptRuleSetsRequest;
(function (ListReceiptRuleSetsRequest) {
    /**
     * @internal
     */
    ListReceiptRuleSetsRequest.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(ListReceiptRuleSetsRequest || (ListReceiptRuleSetsRequest = {}));
export var ListReceiptRuleSetsResponse;
(function (ListReceiptRuleSetsResponse) {
    /**
     * @internal
     */
    ListReceiptRuleSetsResponse.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(ListReceiptRuleSetsResponse || (ListReceiptRuleSetsResponse = {}));
export var ListTemplatesRequest;
(function (ListTemplatesRequest) {
    /**
     * @internal
     */
    ListTemplatesRequest.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(ListTemplatesRequest || (ListTemplatesRequest = {}));
export var TemplateMetadata;
(function (TemplateMetadata) {
    /**
     * @internal
     */
    TemplateMetadata.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(TemplateMetadata || (TemplateMetadata = {}));
export var ListTemplatesResponse;
(function (ListTemplatesResponse) {
    /**
     * @internal
     */
    ListTemplatesResponse.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(ListTemplatesResponse || (ListTemplatesResponse = {}));
export var ListVerifiedEmailAddressesResponse;
(function (ListVerifiedEmailAddressesResponse) {
    /**
     * @internal
     */
    ListVerifiedEmailAddressesResponse.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(ListVerifiedEmailAddressesResponse || (ListVerifiedEmailAddressesResponse = {}));
export var MailFromDomainNotVerifiedException;
(function (MailFromDomainNotVerifiedException) {
    /**
     * @internal
     */
    MailFromDomainNotVerifiedException.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(MailFromDomainNotVerifiedException || (MailFromDomainNotVerifiedException = {}));
export var Message;
(function (Message) {
    /**
     * @internal
     */
    Message.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(Message || (Message = {}));
export var MessageDsn;
(function (MessageDsn) {
    /**
     * @internal
     */
    MessageDsn.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(MessageDsn || (MessageDsn = {}));
export var MessageRejected;
(function (MessageRejected) {
    /**
     * @internal
     */
    MessageRejected.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(MessageRejected || (MessageRejected = {}));
export var MissingRenderingAttributeException;
(function (MissingRenderingAttributeException) {
    /**
     * @internal
     */
    MissingRenderingAttributeException.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(MissingRenderingAttributeException || (MissingRenderingAttributeException = {}));
export var ProductionAccessNotGrantedException;
(function (ProductionAccessNotGrantedException) {
    /**
     * @internal
     */
    ProductionAccessNotGrantedException.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(ProductionAccessNotGrantedException || (ProductionAccessNotGrantedException = {}));
export var PutConfigurationSetDeliveryOptionsRequest;
(function (PutConfigurationSetDeliveryOptionsRequest) {
    /**
     * @internal
     */
    PutConfigurationSetDeliveryOptionsRequest.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(PutConfigurationSetDeliveryOptionsRequest || (PutConfigurationSetDeliveryOptionsRequest = {}));
export var PutConfigurationSetDeliveryOptionsResponse;
(function (PutConfigurationSetDeliveryOptionsResponse) {
    /**
     * @internal
     */
    PutConfigurationSetDeliveryOptionsResponse.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(PutConfigurationSetDeliveryOptionsResponse || (PutConfigurationSetDeliveryOptionsResponse = {}));
export var PutIdentityPolicyRequest;
(function (PutIdentityPolicyRequest) {
    /**
     * @internal
     */
    PutIdentityPolicyRequest.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(PutIdentityPolicyRequest || (PutIdentityPolicyRequest = {}));
export var PutIdentityPolicyResponse;
(function (PutIdentityPolicyResponse) {
    /**
     * @internal
     */
    PutIdentityPolicyResponse.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(PutIdentityPolicyResponse || (PutIdentityPolicyResponse = {}));
export var RawMessage;
(function (RawMessage) {
    /**
     * @internal
     */
    RawMessage.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(RawMessage || (RawMessage = {}));
export var ReorderReceiptRuleSetRequest;
(function (ReorderReceiptRuleSetRequest) {
    /**
     * @internal
     */
    ReorderReceiptRuleSetRequest.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(ReorderReceiptRuleSetRequest || (ReorderReceiptRuleSetRequest = {}));
export var ReorderReceiptRuleSetResponse;
(function (ReorderReceiptRuleSetResponse) {
    /**
     * @internal
     */
    ReorderReceiptRuleSetResponse.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(ReorderReceiptRuleSetResponse || (ReorderReceiptRuleSetResponse = {}));
export var SendBounceRequest;
(function (SendBounceRequest) {
    /**
     * @internal
     */
    SendBounceRequest.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(SendBounceRequest || (SendBounceRequest = {}));
export var SendBounceResponse;
(function (SendBounceResponse) {
    /**
     * @internal
     */
    SendBounceResponse.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(SendBounceResponse || (SendBounceResponse = {}));
export var SendBulkTemplatedEmailRequest;
(function (SendBulkTemplatedEmailRequest) {
    /**
     * @internal
     */
    SendBulkTemplatedEmailRequest.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(SendBulkTemplatedEmailRequest || (SendBulkTemplatedEmailRequest = {}));
export var SendBulkTemplatedEmailResponse;
(function (SendBulkTemplatedEmailResponse) {
    /**
     * @internal
     */
    SendBulkTemplatedEmailResponse.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(SendBulkTemplatedEmailResponse || (SendBulkTemplatedEmailResponse = {}));
export var SendCustomVerificationEmailRequest;
(function (SendCustomVerificationEmailRequest) {
    /**
     * @internal
     */
    SendCustomVerificationEmailRequest.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(SendCustomVerificationEmailRequest || (SendCustomVerificationEmailRequest = {}));
export var SendCustomVerificationEmailResponse;
(function (SendCustomVerificationEmailResponse) {
    /**
     * @internal
     */
    SendCustomVerificationEmailResponse.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(SendCustomVerificationEmailResponse || (SendCustomVerificationEmailResponse = {}));
export var SendEmailRequest;
(function (SendEmailRequest) {
    /**
     * @internal
     */
    SendEmailRequest.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(SendEmailRequest || (SendEmailRequest = {}));
export var SendEmailResponse;
(function (SendEmailResponse) {
    /**
     * @internal
     */
    SendEmailResponse.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(SendEmailResponse || (SendEmailResponse = {}));
export var SendRawEmailRequest;
(function (SendRawEmailRequest) {
    /**
     * @internal
     */
    SendRawEmailRequest.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(SendRawEmailRequest || (SendRawEmailRequest = {}));
export var SendRawEmailResponse;
(function (SendRawEmailResponse) {
    /**
     * @internal
     */
    SendRawEmailResponse.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(SendRawEmailResponse || (SendRawEmailResponse = {}));
export var SendTemplatedEmailRequest;
(function (SendTemplatedEmailRequest) {
    /**
     * @internal
     */
    SendTemplatedEmailRequest.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(SendTemplatedEmailRequest || (SendTemplatedEmailRequest = {}));
export var SendTemplatedEmailResponse;
(function (SendTemplatedEmailResponse) {
    /**
     * @internal
     */
    SendTemplatedEmailResponse.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(SendTemplatedEmailResponse || (SendTemplatedEmailResponse = {}));
export var SetActiveReceiptRuleSetRequest;
(function (SetActiveReceiptRuleSetRequest) {
    /**
     * @internal
     */
    SetActiveReceiptRuleSetRequest.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(SetActiveReceiptRuleSetRequest || (SetActiveReceiptRuleSetRequest = {}));
export var SetActiveReceiptRuleSetResponse;
(function (SetActiveReceiptRuleSetResponse) {
    /**
     * @internal
     */
    SetActiveReceiptRuleSetResponse.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(SetActiveReceiptRuleSetResponse || (SetActiveReceiptRuleSetResponse = {}));
export var SetIdentityDkimEnabledRequest;
(function (SetIdentityDkimEnabledRequest) {
    /**
     * @internal
     */
    SetIdentityDkimEnabledRequest.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(SetIdentityDkimEnabledRequest || (SetIdentityDkimEnabledRequest = {}));
export var SetIdentityDkimEnabledResponse;
(function (SetIdentityDkimEnabledResponse) {
    /**
     * @internal
     */
    SetIdentityDkimEnabledResponse.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(SetIdentityDkimEnabledResponse || (SetIdentityDkimEnabledResponse = {}));
export var SetIdentityFeedbackForwardingEnabledRequest;
(function (SetIdentityFeedbackForwardingEnabledRequest) {
    /**
     * @internal
     */
    SetIdentityFeedbackForwardingEnabledRequest.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(SetIdentityFeedbackForwardingEnabledRequest || (SetIdentityFeedbackForwardingEnabledRequest = {}));
export var SetIdentityFeedbackForwardingEnabledResponse;
(function (SetIdentityFeedbackForwardingEnabledResponse) {
    /**
     * @internal
     */
    SetIdentityFeedbackForwardingEnabledResponse.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(SetIdentityFeedbackForwardingEnabledResponse || (SetIdentityFeedbackForwardingEnabledResponse = {}));
export var SetIdentityHeadersInNotificationsEnabledRequest;
(function (SetIdentityHeadersInNotificationsEnabledRequest) {
    /**
     * @internal
     */
    SetIdentityHeadersInNotificationsEnabledRequest.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(SetIdentityHeadersInNotificationsEnabledRequest || (SetIdentityHeadersInNotificationsEnabledRequest = {}));
export var SetIdentityHeadersInNotificationsEnabledResponse;
(function (SetIdentityHeadersInNotificationsEnabledResponse) {
    /**
     * @internal
     */
    SetIdentityHeadersInNotificationsEnabledResponse.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(SetIdentityHeadersInNotificationsEnabledResponse || (SetIdentityHeadersInNotificationsEnabledResponse = {}));
export var SetIdentityMailFromDomainRequest;
(function (SetIdentityMailFromDomainRequest) {
    /**
     * @internal
     */
    SetIdentityMailFromDomainRequest.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(SetIdentityMailFromDomainRequest || (SetIdentityMailFromDomainRequest = {}));
export var SetIdentityMailFromDomainResponse;
(function (SetIdentityMailFromDomainResponse) {
    /**
     * @internal
     */
    SetIdentityMailFromDomainResponse.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(SetIdentityMailFromDomainResponse || (SetIdentityMailFromDomainResponse = {}));
export var SetIdentityNotificationTopicRequest;
(function (SetIdentityNotificationTopicRequest) {
    /**
     * @internal
     */
    SetIdentityNotificationTopicRequest.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(SetIdentityNotificationTopicRequest || (SetIdentityNotificationTopicRequest = {}));
export var SetIdentityNotificationTopicResponse;
(function (SetIdentityNotificationTopicResponse) {
    /**
     * @internal
     */
    SetIdentityNotificationTopicResponse.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(SetIdentityNotificationTopicResponse || (SetIdentityNotificationTopicResponse = {}));
export var SetReceiptRulePositionRequest;
(function (SetReceiptRulePositionRequest) {
    /**
     * @internal
     */
    SetReceiptRulePositionRequest.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(SetReceiptRulePositionRequest || (SetReceiptRulePositionRequest = {}));
export var SetReceiptRulePositionResponse;
(function (SetReceiptRulePositionResponse) {
    /**
     * @internal
     */
    SetReceiptRulePositionResponse.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(SetReceiptRulePositionResponse || (SetReceiptRulePositionResponse = {}));
export var TestRenderTemplateRequest;
(function (TestRenderTemplateRequest) {
    /**
     * @internal
     */
    TestRenderTemplateRequest.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(TestRenderTemplateRequest || (TestRenderTemplateRequest = {}));
export var TestRenderTemplateResponse;
(function (TestRenderTemplateResponse) {
    /**
     * @internal
     */
    TestRenderTemplateResponse.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(TestRenderTemplateResponse || (TestRenderTemplateResponse = {}));
export var UpdateAccountSendingEnabledRequest;
(function (UpdateAccountSendingEnabledRequest) {
    /**
     * @internal
     */
    UpdateAccountSendingEnabledRequest.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(UpdateAccountSendingEnabledRequest || (UpdateAccountSendingEnabledRequest = {}));
export var UpdateConfigurationSetEventDestinationRequest;
(function (UpdateConfigurationSetEventDestinationRequest) {
    /**
     * @internal
     */
    UpdateConfigurationSetEventDestinationRequest.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(UpdateConfigurationSetEventDestinationRequest || (UpdateConfigurationSetEventDestinationRequest = {}));
export var UpdateConfigurationSetEventDestinationResponse;
(function (UpdateConfigurationSetEventDestinationResponse) {
    /**
     * @internal
     */
    UpdateConfigurationSetEventDestinationResponse.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(UpdateConfigurationSetEventDestinationResponse || (UpdateConfigurationSetEventDestinationResponse = {}));
export var UpdateConfigurationSetReputationMetricsEnabledRequest;
(function (UpdateConfigurationSetReputationMetricsEnabledRequest) {
    /**
     * @internal
     */
    UpdateConfigurationSetReputationMetricsEnabledRequest.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(UpdateConfigurationSetReputationMetricsEnabledRequest || (UpdateConfigurationSetReputationMetricsEnabledRequest = {}));
export var UpdateConfigurationSetSendingEnabledRequest;
(function (UpdateConfigurationSetSendingEnabledRequest) {
    /**
     * @internal
     */
    UpdateConfigurationSetSendingEnabledRequest.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(UpdateConfigurationSetSendingEnabledRequest || (UpdateConfigurationSetSendingEnabledRequest = {}));
export var UpdateConfigurationSetTrackingOptionsRequest;
(function (UpdateConfigurationSetTrackingOptionsRequest) {
    /**
     * @internal
     */
    UpdateConfigurationSetTrackingOptionsRequest.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(UpdateConfigurationSetTrackingOptionsRequest || (UpdateConfigurationSetTrackingOptionsRequest = {}));
export var UpdateConfigurationSetTrackingOptionsResponse;
(function (UpdateConfigurationSetTrackingOptionsResponse) {
    /**
     * @internal
     */
    UpdateConfigurationSetTrackingOptionsResponse.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(UpdateConfigurationSetTrackingOptionsResponse || (UpdateConfigurationSetTrackingOptionsResponse = {}));
export var UpdateCustomVerificationEmailTemplateRequest;
(function (UpdateCustomVerificationEmailTemplateRequest) {
    /**
     * @internal
     */
    UpdateCustomVerificationEmailTemplateRequest.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(UpdateCustomVerificationEmailTemplateRequest || (UpdateCustomVerificationEmailTemplateRequest = {}));
export var UpdateReceiptRuleRequest;
(function (UpdateReceiptRuleRequest) {
    /**
     * @internal
     */
    UpdateReceiptRuleRequest.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(UpdateReceiptRuleRequest || (UpdateReceiptRuleRequest = {}));
export var UpdateReceiptRuleResponse;
(function (UpdateReceiptRuleResponse) {
    /**
     * @internal
     */
    UpdateReceiptRuleResponse.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(UpdateReceiptRuleResponse || (UpdateReceiptRuleResponse = {}));
export var UpdateTemplateRequest;
(function (UpdateTemplateRequest) {
    /**
     * @internal
     */
    UpdateTemplateRequest.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(UpdateTemplateRequest || (UpdateTemplateRequest = {}));
export var UpdateTemplateResponse;
(function (UpdateTemplateResponse) {
    /**
     * @internal
     */
    UpdateTemplateResponse.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(UpdateTemplateResponse || (UpdateTemplateResponse = {}));
export var VerifyDomainDkimRequest;
(function (VerifyDomainDkimRequest) {
    /**
     * @internal
     */
    VerifyDomainDkimRequest.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(VerifyDomainDkimRequest || (VerifyDomainDkimRequest = {}));
export var VerifyDomainDkimResponse;
(function (VerifyDomainDkimResponse) {
    /**
     * @internal
     */
    VerifyDomainDkimResponse.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(VerifyDomainDkimResponse || (VerifyDomainDkimResponse = {}));
export var VerifyDomainIdentityRequest;
(function (VerifyDomainIdentityRequest) {
    /**
     * @internal
     */
    VerifyDomainIdentityRequest.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(VerifyDomainIdentityRequest || (VerifyDomainIdentityRequest = {}));
export var VerifyDomainIdentityResponse;
(function (VerifyDomainIdentityResponse) {
    /**
     * @internal
     */
    VerifyDomainIdentityResponse.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(VerifyDomainIdentityResponse || (VerifyDomainIdentityResponse = {}));
export var VerifyEmailAddressRequest;
(function (VerifyEmailAddressRequest) {
    /**
     * @internal
     */
    VerifyEmailAddressRequest.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(VerifyEmailAddressRequest || (VerifyEmailAddressRequest = {}));
export var VerifyEmailIdentityRequest;
(function (VerifyEmailIdentityRequest) {
    /**
     * @internal
     */
    VerifyEmailIdentityRequest.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(VerifyEmailIdentityRequest || (VerifyEmailIdentityRequest = {}));
export var VerifyEmailIdentityResponse;
(function (VerifyEmailIdentityResponse) {
    /**
     * @internal
     */
    VerifyEmailIdentityResponse.filterSensitiveLog = function (obj) { return (__assign({}, obj)); };
})(VerifyEmailIdentityResponse || (VerifyEmailIdentityResponse = {}));
