// @aws-sdk/client-ses/dist/es/commands/SetActiveReceiptRuleSetCommand.js
import { __extends } from "tslib";
import { SetActiveReceiptRuleSetRequest, SetActiveReceiptRuleSetResponse } from "../models/models_0";
import { deserializeAws_querySetActiveReceiptRuleSetCommand, serializeAws_querySetActiveReceiptRuleSetCommand, } from "../protocols/Aws_query";
import { getSerdePlugin } from "@aws-sdk/middleware-serde";
import { Command as $Command } from "@aws-sdk/smithy-client";
/**
 * <p>Sets the specified receipt rule set as the active receipt rule set.</p>
 *         <note>
 *             <p>To disable your email-receiving through Amazon SES completely, you can call this API
 *                 with RuleSetName set to null.</p>
 *         </note>
 *         <p>For information about managing receipt rule sets, see the <a href="https://docs.aws.amazon.com/ses/latest/DeveloperGuide/receiving-email-managing-receipt-rule-sets.html">Amazon SES Developer Guide</a>.</p>
 *         <p>You can execute this operation no more than once per second.</p>
 */
var SetActiveReceiptRuleSetCommand = /** @class */ (function (_super) {
    __extends(SetActiveReceiptRuleSetCommand, _super);
    // Start section: command_properties
    // End section: command_properties
    function SetActiveReceiptRuleSetCommand(input) {
        var _this = 
        // Start section: command_constructor
        _super.call(this) || this;
        _this.input = input;
        return _this;
        // End section: command_constructor
    }
    /**
     * @internal
     */
    SetActiveReceiptRuleSetCommand.prototype.resolveMiddleware = function (clientStack, configuration, options) {
        this.middlewareStack.use(getSerdePlugin(configuration, this.serialize, this.deserialize));
        var stack = clientStack.concat(this.middlewareStack);
        var logger = configuration.logger;
        var clientName = "SESClient";
        var commandName = "SetActiveReceiptRuleSetCommand";
        var handlerExecutionContext = {
            logger: logger,
            clientName: clientName,
            commandName: commandName,
            inputFilterSensitiveLog: SetActiveReceiptRuleSetRequest.filterSensitiveLog,
            outputFilterSensitiveLog: SetActiveReceiptRuleSetResponse.filterSensitiveLog,
        };
        var requestHandler = configuration.requestHandler;
        return stack.resolve(function (request) {
            return requestHandler.handle(request.request, options || {});
        }, handlerExecutionContext);
    };
    SetActiveReceiptRuleSetCommand.prototype.serialize = function (input, context) {
        return serializeAws_querySetActiveReceiptRuleSetCommand(input, context);
    };
    SetActiveReceiptRuleSetCommand.prototype.deserialize = function (output, context) {
        return deserializeAws_querySetActiveReceiptRuleSetCommand(output, context);
    };
    return SetActiveReceiptRuleSetCommand;
}($Command));
export { SetActiveReceiptRuleSetCommand };
